colossalai/shardformer/layer/utils.py

```diff
@@ -289,3 +289,42 @@ def create_randomizer_with_offset(
         Randomizer.increment_index()
 
     return Randomizer(seed=base_seed)
+
+
+def zigzag_split_batch(batch: Dict[str, torch.Tensor], sp_group):
+    """
+    Split the input along the sequence dimension for Ring Attention. As naively spliting sequence
+    in the causual setting will result in the first ranks having much less workload than the last ranks,
+    we split after "folding" the 2D attention mask in half (https://github.com/zhuzilin/ring-flash-attention/issues/2).
+    For example, for sp_size = 4 and seq_len = 8, we get | s0, s7 | s1, s6 | s2, s5 | s3, s4 |.
+
+    Args:
+        batch (Dict[torch.Tensor]): The input tensors to split.
+        sp_group (ProcessGroup): The process group for sequence parallelism.
+
+    """
+    sp_size = dist.get_world_size(sp_group)
+    sp_rank = dist.get_rank(sp_group)
+    if sp_size > 1:
+        for key, tensor in batch.items():
+            seq_dim = 1 if key != "attention_mask" else 2
```

Comment:
It seems that this function is called after embedding layer. But the seq_dim is not always 1 for hidden states. E.g. for chatglm, its seq_dim is 0.
![image](https://github.com/user-attachments/assets/dc38d834-7631-40b5-aa26-030dfa311dd0)

Reply:
Now provides a seq_dim argument.